core/Pos/Binary/Crypto.hs

```diff
@@ -43,263 +34,192 @@ import           Pos.Crypto.Signing         (ProxyCert (..), ProxySecretKey (..)
-constantSizedBinaryToStoreGet :: Binary.Binary a => Int -> Store.Peek a
-constantSizedBinaryToStoreGet bytes = do
-    x <- getBytes bytes
-    case Binary.decodeOrFail (BSL.fromStrict x) of
-        Left (_, _, err) -> fail err
-        Right (bs, _, res)
-            | BSL.null bs -> pure res
-            | otherwise   -> fail "unconsumed input"
-
--- [CSL-1122] TODO: more efficient 'put' and 'get' are possible if 'Store'
--- instances are added into pvss-haskell
-#define BiPvss(T, PT, BYTES) \
+#define BiPvss(T, PT) \
   instance Bi T where {\
-    size = ConstSize BYTES ;\
-    put = labelP "T" . putBytes . BSL.toStrict . Binary.encode ;\
-    get = label "T" $ constantSizedBinaryToStoreGet BYTES };\
+    encode = encodeBinary ;\
+    decode = decodeBinary };\
   deriving instance Bi PT ;\
 
-BiPvss (Pvss.PublicKey, VssPublicKey, 33)    -- yes it's 33 and not 32
-BiPvss (Pvss.KeyPair, VssKeyPair, 65)        -- 32+33
-BiPvss (Pvss.Secret, Secret, 33)
-BiPvss (Pvss.DecryptedShare, Share, 101)     -- 4+33+64
-BiPvss (Pvss.EncryptedShare, EncShare, 101)
-BiPvss (Pvss.Proof, SecretProof, 64)
-
-instance Store.Store Pvss.ExtraGen where
-    size = ConstSize 33
-    poke = labelP "Pvss.ExtraGen" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.ExtraGen" $ constantSizedBinaryToStoreGet 33
-instance Store.Store Pvss.Commitment where
-    size = ConstSize 33
-    poke = labelP "Pvss.Commitment" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.Commitment" $ constantSizedBinaryToStoreGet 33
-
-Store.makeStore ''SecretSharingExtra
+BiPvss (Pvss.PublicKey, VssPublicKey)
+BiPvss (Pvss.KeyPair, VssKeyPair)
+BiPvss (Pvss.Secret, Secret)
+BiPvss (Pvss.DecryptedShare, Share)
+BiPvss (Pvss.EncryptedShare, EncShare)
+BiPvss (Pvss.Proof, SecretProof)
+
+instance Bi Pvss.ExtraGen where
+    encode = encodeBinary
+    decode = decodeBinary
+
+instance Bi Pvss.Commitment where
+    encode = encodeBinary
+    decode = decodeBinary
+
 instance Bi SecretSharingExtra where
-    put = labelP "SecretSharingExtra" . Store.poke
-    get = label "SecretSharingExtra" $ Store.peek
-    size = Store.size
+    encode (SecretSharingExtra eg comms) = encodeListLen 2
+                                        <> encode eg
+                                        <> encode comms
+    decode = SecretSharingExtra
+          <$  enforceSize "SecretSharingExtra" 2
+          <*> decode
+          <*> decode
 
 deriving instance Bi (AsBinary SecretSharingExtra)
 
 ----------------------------------------------------------------------------
 -- SecretSharing AsBinary
 ----------------------------------------------------------------------------
 
-#define BiMacro(B, BYTES) \
+#define BiMacro(B) \
   instance Bi (AsBinary B) where {\
-    size = ConstSize BYTES ;\
-    put (AsBinary bs) = putBytes bs ;\
-    get = label "B (BYTES bytes)" $ AsBinary <$> getBytes BYTES}; \
+    encode (AsBinary bs) = encode bs ;\
+    decode = AsBinary <$> decode}; \
```

Comment:
1. Regarding bug: if we decode `AsBinary VssPublicKey` which internally stores a `ByteString` with length 100, it will be decoded successfully and `AsBinary` will keep 100 bytes instead. When we later call `fromBinary`, it will fail. But we want the failure to happen when we decode bytes into `AsBinary VssPublicKey`, not when we extract `VssPublicKey` from `AsBinary VssPublicKey`.
Our code can't create invalid `AsBinary VssPublicKey`, but we can receive malformed data from the network.
2. Regarding overhead: yes, multiple `AsBinary` don't increase the overhead. But one `AsBinary` adds a slight overhead relative to no `AsBinary`. And previously it wasn't the case. I am quite sure that length of `serialize (asBinary randomVssPublicKey)` will be 37 while length of `serialize randomVssPublicKey` is 35 (sorry, I am too lazy to check it).
Ideally serialization of types with `AsBinary` should be the same as of the same types without `AsBinary`. But the overhead is small and probably doesn't worth a discussion.

P. S. You can ignore it for now until the PR is merged, I agree that merging it has higher priority 😉